File: ruby/23.rb
def divisors(n)
  d = (2..Math.sqrt(n)).select { |q| n%q == 0 }
  d = d + d.map { |q| n / q } + [1]
  d.uniq
end

def sigma(n)
  divisors(n).inject(&:+)
end

limit = 28123
abundants = (1..limit).select { |n| sigma(n) > n }
numbers = (0..limit).to_a

(0...abundants.length).each do |i|
  (i...abundants.length).each do |j|
    index = abundants[i] + abundants[j]
    numbers[index] = 0 if index <= limit
  end
end

puts numbers.inject(&:+)

File: scala/23.scala
import scala.math._

def divisors(n: Int) = {
  val d = (2 to sqrt(n).toInt).filter(n % _ == 0)
  Set() ++ d ++ d.map(n / _) ++ Set(1)
}

def sigma(n: Int) = divisors(n).reduceLeft(_ + _)

val limit = 28123
val abundants = (1 to limit).filter(n => sigma(n) > n).toArray
val numbers = (0 to limit).toArray

for (
  i <- 0 until abundants.length;
  j <- i until abundants.length
) {
  val index = abundants(i) + abundants(j)
  if (index <= limit)
    numbers(index) = 0
}

println(numbers.reduceLeft(_ + _))
